catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java

```diff
@@ -403,7 +410,11 @@ private ContentItem readContent(URI uri) throws StorageException {
     if (REF_EXT.equals(extension)) {
       extension = FilenameUtils.getExtension(FilenameUtils.removeExtension(filename));
       try {
-        reference = new URI(new String(Files.readAllBytes(file), Charset.forName("UTF-8")));
+        reference =
+            new URI(
```

Comment:
❓ Wait...so the URI is the _contents_ of the file? That can't be correct, can it? Or are the contents of the file a uri-able string?

Reply:
I got confused on this also. Like 5 times now lol.

I'm probably going to pair with someone to see what is going on and try to refactor it.

Reply:
After looking at this some more, I figured out what is going on.

[There are times that we want to store an "external reference" to something that is already stored](https://github.com/codice/ddf/pull/4220/files/03bdb890dbae452b8439f679772e0ee6516662f5#diff-9b68a9cb68bedc7e0609dcfdd52cf3dfR587). This "external reference"'s file content is just the url of the file we eventually want to read.

So when we are reading the content here, we check to see if we are given an "external reference" (a file containing a URL of the real data we want to read), and if we are, we read the file content of that "external reference" and [prepare to open up the stream to the real data below](https://github.com/codice/ddf/pull/4220/files/03bdb890dbae452b8439f679772e0ee6516662f5#diff-9b68a9cb68bedc7e0609dcfdd52cf3dfR441).